### PR TITLE
[DependencyInjection] Optimize exporting variables

### PR DESCRIPTION
**Pull request by svandervlugt**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no  
| Deprecations? | no 
| Tests pass?   | yes   
| Fixed tickets | #28292 
| License       | MIT

Store previously resolved variables so that the relatively expensive function [resolveEnvPlaceholders](https://github.com/symfony/symfony/blob/3e7b029524bf724abfd603c17795af39bb289615/src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php#L1776) is not called unnecessarily.

The effect on running ```bin/console``` with an empty cache in a big application (200+ env variables) in seconds:

| Before     | After      |
|------------|------------|
| 24.1       | 15.8       |
| 24.5       | 16.0       |
| 24.7       | 16.3       |
| 24.2       | 16.0       |
|  Avg: 24.4 | Avg:16.0   |

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php [hunk 1]
```diff
@@ -1765,6 +1766,9 @@ private function export($value)
 
     private function doExport($value, $resolveEnv = false)
     {
+        if ($resolveEnv && \is_string($value) && array_key_exists($value, $this->exportedVariables)) {
```
Comment (ostrolucky) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: result of `$resolveEnv && \is_string($value)` can be saved into variable so it doesn't need to be evaluated later again

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php [hunk 2]
```diff
@@ -1786,6 +1790,10 @@ private function doExport($value, $resolveEnv = false)
             }
         }
 
+        if ($resolveEnv && \is_string($value) && \is_string($export)) {
```
Comment (ostrolucky) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: There doesn't seem to be reason to use `\is_string($export)`

**Review by ostrolucky — CHANGES_REQUESTED** [review]

**Review by nicolas-grekas — CHANGES_REQUESTED** [review]
Thanks for the PR, I confirm a 5% improvement (and a 5% peak memory increase on my case)
in addition to @ostrolucky's comment, we should also empty the new property when leaving the dump function.

**Comment by svandervlugt** [comment]
I don't know why 2 of the 3 travis CI builds suddenly fail. I'm not sure whether this is related to my changes as they are in the Lock-component and the Translator in the FrameworkBundle.

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php [hunk 3]
```diff
@@ -294,6 +295,7 @@ public function dump(array $options = array())
         $this->inlinedRequires = array();
         $this->circularReferences = array();
         $this->locatedIds = array();
+        $this->exportedVariables = array();
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: This should be duplicated L127 above

**Comment by svandervlugt** [comment]
Another failing build. Still not convinced it is my changes.
appveyor was successful before my last commit, now fails with  ```..KO src\Symfony/Bundle/SecurityBundle```. and no further error.

One of the three travis-builds fails on the FrameworkBundle, Translation component and Validator component. Before each of these a ```  - Installing symfony/dependency-injection (dev-master 88f240c): Loading from cache``` is done. Doesn't this effectively overwrite my changes?

**Review by ostrolucky — APPROVED** [review]

**Comment by svandervlugt** [comment]
@nicolas-grekas the PR has been updated with the requested changes. 
After the last rebase all checks have passed.

**Review by nicolas-grekas — APPROVED** [review]

**Review by dunglas — APPROVED** [review]

**Pull request merged**